neps/nep-0384.md

```diff
@@ -0,0 +1,108 @@
+    abortTransactionUnless(connection.verifyConnectionState(proofHeight, proofInit, counterpartyConnectionIdentifier, expectedConnectionEnd))
+    abortTransactionUnless(connection.verifyClientState(proofHeight, proofClient, clientState))
+    abortTransactionUnless(connection.verifyClientConsensusState(
+      proofHeight, proofConsensus, counterpartyClientIdentifier, consensusHeight, expectedConsensusState))
+
+    provableStore.set(connectionPath(identifier), connection)
+    addConnectionToClient(clientIdentifier, identifier)
+}
+
+// ConnOpenAck relays acceptance of a connection open attempt from chain B back to chain A (this code is executed on chain A).
+function connOpenAck(
+  identifier: Identifier,
+  clientState: ClientState,
+  version: string,
+  counterpartyIdentifier: Identifier,
+  proofTry: CommitmentProof,
+  proofClient: CommitmentProof,
+  proofConsensus: CommitmentProof,
+  proofHeight: Height,
+  consensusHeight: Height) {
+    abortTransactionUnless(consensusHeight < getCurrentHeight())
+    abortTransactionUnless(validateSelfClient(clientState))
+    connection = provableStore.get(connectionPath(identifier))
+    abortTransactionUnless((connection.state === INIT && connection.version.indexOf(version) !== -1)
+    expectedConsensusState = getHostConsensusState(consensusHeight) // we need to return a near header at this height.
+    expectedConnectionEnd = ConnectionEnd{TRYOPEN, identifier, getCommitmentPrefix(),
+                             connection.counterpartyClientIdentifier, connection.clientIdentifier,
+                             version, connection.delayPeriodTime, connection.delayPeriodBlocks}
+    abortTransactionUnless(connection.verifyConnectionState(proofHeight, proofTry, counterpartyIdentifier, expectedConnectionEnd)) // so we can verify that the counterparty chain has the correct header.
+    abortTransactionUnless(connection.verifyClientState(proofHeight, proofClient, clientState))
+    abortTransactionUnless(connection.verifyClientConsensusState(
+      proofHeight, proofConsensus, connection.counterpartyClientIdentifier, consensusHeight, expectedConsensusState))
+    connection.state = OPEN
+    connection.version = version
+    connection.counterpartyConnectionIdentifier = counterpartyIdentifier
+    provableStore.set(connectionPath(identifier), connection)
+}
+```
+
+The api `env::block_hash` allows us validate the full header submitted via contract calldata by comparing hashes.
+
+## Rationale and alternatives
+
+A few other designs were considered such as a function `env::block_header` which would return the full header at a given height, but was discouraged against a result of expensive db calls. This mapping of block numbers to headers are cheap enough to be stored on-chain.
+
+
+## Specification
+
+```rust
+/// This function should return a blockhash for 256 of the most reccent blocks
```

Comment:
> 256 of the most reccent blocks

What *exactly* does that mean?

1. every block in the range of `current block height - 256` up to `current block height - 1`?
2. the last 256 blocks that are not missing?
3. the last 256 final blocks?

Option 3 would make the implementation much easier and consequently the host function call cheaper. But I suspect you want 1. or 2.

Reply:
1. Is the intended case

Reply:
ok - but this means that there might be situations where there are no blocks within the last '256' heights (if network is in trouble and cannot gather consensus etc)  - and we're ok with that?